### PR TITLE
[node-metrics] Ability to collect system metrics from Aptos nodes

### PR DESCRIPTION
**Pull request by ibalajiarun**
### Description

This PR introduces the ability to collect _platform-agnostic_ system metrics from Aptos nodes and expose them as prometheus metrics. The metrics are also exported to Aptos via telemetry service. There are multiple collectors to collect various categories of system metrics (CPU, memory, disk, etc) using the `sysinfo` crate. 

### Test Plan

Added unit tests 

Below is the output from local testing:
```
# HELP node_disk_available_space Total available disk size in bytes
# TYPE node_disk_available_space counter
node_disk_available_space{file_system="apfs",name="Data",type="SSD"} 116468510720
# HELP node_disk_total_space Total disk size in bytes
# TYPE node_disk_total_space counter
node_disk_total_space{file_system="apfs",name="Data",type="SSD"} 994662584320
# HELP node_loadavg_load1 1m load average.
# TYPE node_loadavg_load1 gauge
node_loadavg_load1 25.810546875
# HELP node_loadavg_load15 15m load average.
# TYPE node_loadavg_load15 gauge
node_loadavg_load15 6.31591796875
# HELP node_loadavg_load5 5m load average.
# TYPE node_loadavg_load5 gauge
node_loadavg_load5 10.93896484375
# HELP node_network_total_packets_received Total number of incoming packets
# TYPE node_network_total_packets_received counter
node_network_total_packets_received{interface_name="en0"} 4476845
# HELP node_network_total_packets_transmitted Total number of outgoing packets
# TYPE node_network_total_packets_transmitted counter
node_network_total_packets_transmitted{interface_name="en0"} 1953492
# HELP node_network_total_received Total number of received bytes
# TYPE node_network_total_received counter
node_network_total_received{interface_name="en0"} 4156922880
# HELP node_network_total_transmitted Total number of transmitted bytes
# TYPE node_network_total_transmitted counter
node_network_total_transmitted{interface_name="en0"} 930886656
# HELP node_process_cpu_usage CPU usage.
# TYPE node_process_cpu_usage gauge
node_process_cpu_usage 781.2310180664063
# HELP node_process_disk_total_read_bytes Total bytes read.
# TYPE node_process_disk_total_read_bytes gauge
node_process_disk_total_read_bytes 6434816
# HELP node_process_disk_total_written_bytes Total bytes written.
# TYPE node_process_disk_total_written_bytes gauge
node_process_disk_total_written_bytes 23715840
# HELP node_process_memory Memory usage in bytes.
# TYPE node_process_memory gauge
node_process_memory 231948
# HELP node_process_run_time Run time of the process in seconds.
# TYPE node_process_run_time gauge
node_process_run_time 30
# HELP node_process_start_time Starts time of the process in seconds since epoch.
# TYPE node_process_start_time gauge
node_process_start_time 1663016819
# HELP node_process_virtual_memory Virtual memory usage in bytes.
# TYPE node_process_virtual_memory gauge
node_process_virtual_memory 420680204
# HELP node_system_cpu_info CPU information.
# TYPE node_system_cpu_info gauge
node_system_cpu_info{brand="Apple M1 Max",vendor="Apple"} 1
# HELP node_system_cpu_usage CPU usage.
# TYPE node_system_cpu_usage gauge
node_system_cpu_usage{cpu_id="0"} 100
node_system_cpu_usage{cpu_id="cpu10_idx10"} 100
node_system_cpu_usage{cpu_id="cpu1_idx1"} 100
node_system_cpu_usage{cpu_id="cpu2_idx2"} 100
node_system_cpu_usage{cpu_id="cpu3_idx3"} 100
node_system_cpu_usage{cpu_id="cpu4_idx4"} 100
node_system_cpu_usage{cpu_id="cpu5_idx5"} 100
node_system_cpu_usage{cpu_id="cpu6_idx6"} 100
node_system_cpu_usage{cpu_id="cpu7_idx7"} 100
node_system_cpu_usage{cpu_id="cpu8_idx8"} 100
node_system_cpu_usage{cpu_id="cpu9_idx9"} 100
# HELP node_system_mem_free Memory free.
# TYPE node_system_mem_free gauge
node_system_mem_free 1588800
# HELP node_system_mem_total Memory total.
# TYPE node_system_mem_total counter
node_system_mem_total 34359738
# HELP node_system_mem_used Memory used.
# TYPE node_system_mem_used gauge
node_system_mem_used 32770938
# HELP node_system_swap_free Swap memory free.
# TYPE node_system_swap_free gauge
node_system_swap_free 0
# HELP node_system_swap_total Swap memory total.
# TYPE node_system_swap_total counter
node_system_swap_total 0
# HELP node_system_swap_used Swap memory used.
# TYPE node_system_swap_used gauge
node_system_swap_used 0
```

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4124)

### CODE REVIEW

**Review by grao1991 — COMMENTED** [review]
Can we have break down for CPU usage?
Can we have both rss and working_set for memory usage?

**Comment by ibalajiarun** [comment]
> Can we have break down for CPU usage? Can we have both rss and working_set for memory usage? Can we have IOPS?

@grao1991 Yes, that will be in a separate PR. It seems we can only collect those explicitly for linux only for now. This PR is for platform agnostic metrics.

**Review by grao1991 — APPROVED** [review]

**Review by JoshLind — APPROVED** [review]
Thanks @ibalajiarun! 😄 Looks reasonable to me!

File: crates/node-metrics/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+authors = ["Aptos Labs <opensource@aptoslabs.com>"]
+description = "Host system prometheus metrics collector"
+repository = "https://github.com/aptos-labs/aptos-core"
+homepage = "https://aptoslabs.com"
+license = "Apache-2.0"
+publish = false
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+aptos-infallible = { path = "../../crates/aptos-infallible" }
```
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply crates/node-metrics/Cargo.toml:14]: We now run `cargo sort` check in CICD, thats why these needs sorted.

File: crates/node-metrics/src/collectors/common.rs [hunk 2]
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+pub const NAMESPACE: &str = "node";
```
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply crates/node-metrics/src/collectors/common.rs:4]: They won't be left alone. There will be some prometheus counters for measuring latency of the collectors. I am working on them in a separate PR.

File: crates/node-metrics/src/collectors/network_collector.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            &self.total_received,
+            &self.total_transmitted,
+            &self.total_packets_received,
+            &self.total_packets_transmitted,
+            &self.total_errors_on_received,
+            &self.total_errors_on_transmitted,
+        ]
+    }
+
+    fn collect(&self) -> Vec<MetricFamily> {
```
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply crates/node-metrics/src/collectors/network_collector.rs:114]: Right now, its the usual cadence of every `15s`, but I have reduced the overhead significantly. `collect()` on all collectos in this crate takes about 100 micro seconds, which I think is acceptable.

File: crates/aptos-metrics-core/src/const_metric.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+impl ConstMetric {
+    pub fn new_counter(
+        desc: Desc,
+        value: f64,
+        label_values: Option<&[String]>,
+    ) -> Result<ConstMetric> {
+        let label_values = label_values.unwrap_or_default();
+        if desc.variable_labels.len() != label_values.len() {
```
Comment (sitalkedia) [inline crates/aptos-metrics-core/src/const_metric.rs:28]: Should you be checking for the actual labels as well?

File: crates/node-metrics/src/collectors/cpu_collector.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+use prometheus::{
+    core::{Collector, Desc, Describer},
+    proto::MetricFamily,
+    Opts,
+};
+use std::sync::Arc;
+use sysinfo::{CpuExt, CpuRefreshKind, RefreshKind, System, SystemExt};
+
+/// A Collector for exposing CPU metrics
+pub(crate) struct CpuCollector {
```
Comment (sitalkedia) [inline crates/node-metrics/src/collectors/cpu_collector.rs:16]: nit: I think a better name would be `CpuMetricsCollector`?

File: crates/node-metrics/src/collectors/disk_collector.rs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+/// A Collector for exposing Disk metrics
+pub(crate) struct DiskCollector {
+    system: Arc<Mutex<System>>,
+
+    total_space: Desc,
+    available_space: Desc,
+}
+
+impl DiskCollector {
```
Comment (sitalkedia) [inline crates/node-metrics/src/collectors/disk_collector.rs:26]: nit `DiskMetricsCollector`

**Review by sitalkedia — APPROVED** [review]
Thanks a lot, @ibalajiarun for collecting these very useful system metrics. Do you know what's the overhead of collecting these metrics and how frequently are we collecting them?

File: crates/aptos-metrics-core/src/const_metric.rs (hunk 4 above)
Comment (ibalajiarun) [reply crates/aptos-metrics-core/src/const_metric.rs:28]: Not unless label_values also includes the label along with the value. The assumption here is that the label_values are in the same order as the labels, which is the standard assumption in the prometheus crate.

File: crates/aptos-telemetry/Cargo.toml [hunk 7]
```diff
@@ -39,6 +39,7 @@ aptos-metrics-core = { path = "../../crates/aptos-metrics-core" }
 aptosdb = { path = "../../storage/aptosdb" }
 consensus = { path = "../../consensus" }
 network = { path = "../../network" }
+node-metrics = { path = "../../crates/node-metrics" }
```
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply crates/aptos-telemetry/Cargo.toml:42]: I used `host-metrics` but we may not be actually collecting host metrics. How about `node-system-metrics`?

**Comment by sitalkedia** [comment]
@ibalajiarun  - As a follow up from this, can you make sure to add/update the grafana dashboards to monitor these newly added metrics?

File: crates/aptos-telemetry/Cargo.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply crates/aptos-telemetry/Cargo.toml]: Actually, `node-resource-metrics` is good.

File: crates/node-metrics/src/collectors/common.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply crates/node-metrics/src/collectors/common.rs]: Added histogram to measure latency of collect operations as well.

File: crates/node-metrics/src/collectors/cpu_collector.rs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    cpu_info: Desc,
+}
+
+impl CpuCollector {
+    fn new() -> Self {
+        let system = Arc::new(Mutex::new(System::new_with_specifics(
+            RefreshKind::new().with_cpu(CpuRefreshKind::everything()),
+        )));
+
+        let cpu = Opts::new("system_cpu_usage", "CPU usage.")
```
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply crates/node-metrics/src/collectors/cpu_collector.rs]: Done.

**Comment by ibalajiarun** [comment]
> Thanks a lot, @ibalajiarun for collecting these very useful system metrics. Do you know what's the overhead of collecting these metrics and how frequently are we collecting them?

@sitalkedia the overhead is 100 micro seconds per collect call as measured in our recommended hardware spec, so I set the default collect frequency of 15 seconds. I've also added latency histogram to monitor the collect calls, which I will add to dashboard. If it seems to hit us, we can tune the collection frequency.

**Pull request merged**